### PR TITLE
filtered list of branches by category name, day and borough

### PR DESCRIPTION
**Pull request by DwinaTech**

### CODE REVIEW

File: backend/src/controllers/get_controller.js [hunk 1]
```diff
@@ -1,17 +1,33 @@
...
+module.exports = {
+  getAllOrgainisation: async () => {
+    const result = await Organisation
+      .query()
+      .eager('[branch, branch.[address, address.[location] service, service.[categories]] ]');
+    return result;
+  },
 
-Model.knex(knex);
+  getListOfBranchesFilteredByCategory: categoryName =>
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: try to shorten method key name `getBranchByCategory`...

File: backend/src/controllers/get_controller.js [hunk 2]
```diff
@@ -1,17 +1,33 @@
...
+    Categories.query().skipUndefined()
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: https://vincit.github.io/objection.js/#eager-loading , eager is better i guess, less response time is preferred.

File: backend/src/controllers/get_controller.js [hunk 3]
```diff
@@ -1,17 +1,33 @@
...
+        .eager('[address, address.[location] service, service.[categories]]')
+        .where('id', data.branch_id)),
 
-const getAllOrgainisation = async () => {
-  const result = await Organisation
-    .query()
-    .eager('[branch, branch.[address, address.[location] service, service.[categories]] ]');
-  return result;
-}
+  getListOfBranchesFilteredByDay: day =>
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: ditto

File: backend/src/controllers/get_controller.js [hunk 4]
```diff
@@ -1,17 +1,33 @@
...
+    Service.query().skipUndefined()
+      .then(services => services.filter(data => data.service_days.includes(day)))
+      .map(serviceData => Branch.query()
+        .eager('[address, address.[location] service, service.[categories]]')
+        .where('id', serviceData.branch_id)),
 
-module.exports = {
-  getAllOrgainisation
+  getListOfBranchesFilteredByBorough: boroughName =>
+    Branch.query().skipUndefined()
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: just try to shorten the method name, nice one

File: backend/src/controllers/postInitialData.js [hunk 5]
```diff
@@ -10,35 +10,35 @@ const loadSeedToDb = async () => {
...
-              long: 'not provided'
-            }]
-          }]
-        }
-      })
+          }
+        ))
     };
-    (async () => transaction(Organisation.knex(), trx => 
+    return async () => transaction(Organisation.knex(), trx =>
```
Comment (sentayhuJo) [inline backend/src/controllers/postInitialData.js]: this function shouldn't return anything , self execute function ()();

File: backend/src/routes/api.js [hunk 6]
```diff
@@ -1,7 +1,12 @@
 
 import { postOrganisation } from '../controllers/post_controller';
-import { getAllOrgainisation } from '../controllers/get_controller';
 import { loadSeedToDb } from '../controllers/postInitialData';
+import {
+  getAllOrgainisation,
+  getListOfBranchesFilteredByCategory,
```
Comment (sentayhuJo) [inline backend/src/routes/api.js]: ditto

File: backend/src/routes/api.js [hunk 7]
```diff
@@ -12,6 +17,20 @@ module.exports = router => {
     getAllOrgainisation().then(data => res.json(data));
   });
   router.get('/service/migrate', (req, res) => {
-    loadSeedToDb().then(data => res.json({ Migration: 'Data migration to database completed successfully !' }));
-  })
+    loadSeedToDb().then(() => res.json({ Migration: 'Data migration to database completed successfully !' }));
+  });
+
+  router.get('/service/category', async (req, res) => {
+    const category = req.query.category.replace(/[-]/g, ' ');
```
Comment (sentayhuJo) [inline backend/src/routes/api.js]: why regex ?
and mind when you send the query parameter form the front end.

File: backend/src/routes/api.js [hunk 8]
```diff
@@ -12,6 +17,20 @@ module.exports = router => {
...
+    getListOfBranchesFilteredByCategory(category).then(data => res.json(data))
+  });
+
+  router.get('/service/days', async (req, res) => {
+    getListOfBranchesFilteredByDay(req.query.day).then(data => res.json(data))
+  });
+
+  router.get('/service/borough', async (req, res) => {
+    const borough = req.query.boroughName.replace(/[-]/g, ' ');
```
Comment (sentayhuJo) [inline backend/src/routes/api.js]: ditto

File: backend/src/routes/api.js (hunk 7 above)
Comment (sentayhuJo) [inline backend/src/routes/api.js]: ditto

File: backend/src/routes/api.js [hunk 9]
```diff
@@ -12,6 +17,20 @@ module.exports = router => {
...
+    loadSeedToDb().then(() => res.json({ Migration: 'Data migration to database completed successfully !' }));
+  });
+
+  router.get('/service/category', async (req, res) => {
+    const category = req.query.category.replace(/[-]/g, ' ');
+    getListOfBranchesFilteredByCategory(category).then(data => res.json(data))
+  });
+
+  router.get('/service/days', async (req, res) => {
+    getListOfBranchesFilteredByDay(req.query.day).then(data => res.json(data))
```
Comment (sentayhuJo) [inline backend/src/routes/api.js]: use variable to hold the incoming query, and be consistent

File: backend/db/seeds/development/shows_seed.js [hunk 10]
```diff
@@ -1,6 +1,6 @@
 const { Model } = require('objection');
 const { transaction } = require('objection');
-const knex =require('../../../config');
+const knex = require('../../../config');
 const Organisation = require('../../../src/model/Organisation');
 const data = require('../../../data.json');
```
[the remark this answers is not in the document]
Comment (DwinaTech) [reply backend/db/seeds/development/shows_seed.js]: I did not do anything here just fixed some lint errors for this file

File: backend/src/controllers/get_controller.js (hunk 1 above)
Comment (DwinaTech) [reply backend/src/controllers/get_controller.js]: Done

File: backend/src/routes/api.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (DwinaTech) [reply backend/src/routes/api.js]: Done

File: backend/src/routes/api.js (hunk 7 above)
Comment (DwinaTech) [reply backend/src/routes/api.js]: Cool, better idea, can be fixed at frontend

File: backend/src/routes/api.js (hunk 8 above)
Comment (DwinaTech) [reply backend/src/routes/api.js]: Fixed

File: backend/src/routes/api.js (hunk 7 above)
Comment (DwinaTech) [reply backend/src/routes/api.js]: Fixed

File: backend/src/routes/api.js (hunk 9 above)
Comment (DwinaTech) [reply backend/src/routes/api.js]: Done

File: backend/src/controllers/get_controller.js (hunk 2 above)
Comment (DwinaTech) [reply backend/src/controllers/get_controller.js]: Done

File: backend/src/routes/api.js (hunk 6 above)
Comment (DwinaTech) [reply backend/src/routes/api.js]: Done

File: backend/src/controllers/postInitialData.js (hunk 5 above)
Comment (DwinaTech) [reply backend/src/controllers/postInitialData.js]: yeah that is true but it is giving error at eslint, it is saying should have return value, but still work in this way also.

File: backend/db/seeds/development/shows_seed.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (sentayhuJo) [reply backend/db/seeds/development/shows_seed.js]: cool

File: backend/src/controllers/get_controller.js [hunk 11]
```diff
@@ -11,23 +11,30 @@ module.exports = {
...
-        .eager('[address, address.[location] service, service.[categories]]')
-        .where('id', data.branch_id)),
+  getBranchByCategory: async categoryName => {
+    const result = await Organisation
+      .query()
+      .eagerAlgorithm(Organisation.JoinEagerAlgorithm)
+      .eager('[branch, branch.[address, address.[location] service, service.[categories]] ]')
+      .where('branch:service:categories.cat_name', 'like', `%${categoryName}%`)
+    return result;
+  },
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: Catch the error.... `try { } catch(e) { }`

File: backend/src/controllers/get_controller.js [hunk 12]
```diff
@@ -11,23 +11,30 @@ module.exports = {
...
-        .eager('[address, address.[location] service, service.[categories]]')
-        .where('id', serviceData.branch_id)),
+  getBranchByDay: async day => {
+    const result = await Organisation
+      .query()
+      .eagerAlgorithm(Organisation.JoinEagerAlgorithm)
+      .eager('[branch, branch.[address, address.[location] service, service.[categories]] ]')
+      .where('branch:service.service_days', 'like', `%${day}%`)
+    return result;
+  },
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: ditto

File: backend/src/controllers/get_controller.js [hunk 13]
```diff
@@ -11,23 +11,30 @@ module.exports = {
...
-      .eager('[address, address.[location] service, service.[categories]]')
-      .where('borough', boroughName)
+  getBranchByBorough: async boroughName => {
+    const result = await Organisation
+      .query()
+      .eagerAlgorithm(Organisation.JoinEagerAlgorithm)
+      .eager('[branch, branch.[address, address.[location] service, service.[categories]] ]')
+      .where('branch.borough', 'like', `%${boroughName}%`)
+    return result;
+  }
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: ditto

File: backend/src/controllers/get_controller.js (hunk 11 above)
Comment (DwinaTech) [reply backend/src/controllers/get_controller.js]: Done

File: backend/src/controllers/get_controller.js (hunk 12 above)
Comment (DwinaTech) [reply backend/src/controllers/get_controller.js]: Done

File: backend/src/controllers/get_controller.js (hunk 13 above)
Comment (DwinaTech) [reply backend/src/controllers/get_controller.js]: Done

**Comment by sentayhuJo** [comment]
merge conflict ....

File: backend/src/controllers/get_controller.js [hunk 14]
```diff
@@ -1,17 +1,56 @@
...
+import Categories from '../model/Categories';
+import Branch from '../model/Branch';
+import Service from '../model/Service';
+
+module.exports = {
+  getAllOrgainisation: async () => {
+    const result = await Organisation
+      .query()
+      .eager('[branch, branch.[address, address.[location] service, service.[categories]] ]');
+    try {
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: nothing to catch this way. wrap the actual query execution.

File: backend/src/controllers/get_controller.js [hunk 15]
```diff
@@ -1,17 +1,56 @@
...
+  },
 
-Model.knex(knex);
+  getBranchByCategory: async categoryName => {
+    const result = await Organisation
+      .query()
+      .eagerAlgorithm(Organisation.JoinEagerAlgorithm)
+      .eager('[branch, branch.[address, address.[location] service, service.[categories]] ]')
+      .where('branch:service:categories.cat_name', 'like', `%${categoryName}%`)
+    try {
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: ditto

File: backend/src/controllers/get_controller.js [hunk 16]
```diff
@@ -1,17 +1,56 @@
...
-    .eager('[branch, branch.[address, address.[location] service, service.[categories]] ]');
-  return result;
-}
+  getBranchByDay: async day => {
+    const result = await Organisation
+      .query()
+      .eagerAlgorithm(Organisation.JoinEagerAlgorithm)
+      .eager('[branch, branch.[address, address.[location] service, service.[categories]] ]')
+      .where('branch:service.service_days', 'like', `%${day}%`)
+    try {
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: ditto

File: backend/src/controllers/get_controller.js [hunk 17]
```diff
@@ -1,17 +1,56 @@
...
 
-module.exports = {
-  getAllOrgainisation
+  getBranchByBorough: async boroughName => {
+    const result = await Organisation
+      .query()
+      .eagerAlgorithm(Organisation.JoinEagerAlgorithm)
+      .eager('[branch, branch.[address, address.[location] service, service.[categories]] ]')
+      .where('branch.borough', 'like', `%${boroughName}%`)
+    try {
```
Comment (sentayhuJo) [inline backend/src/controllers/get_controller.js]: ditto

**Pull request merged**